File: lib/python/src/main/opentoken/tokens/token_registry.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import Dict, List
from opentoken.attributes.attribute_expression import AttributeExpression
from opentoken.tokens.definitions.t1_token import T1Token
from opentoken.tokens.definitions.t2_token import T2Token
from opentoken.tokens.definitions.t3_token import T3Token
from opentoken.tokens.definitions.t4_token import T4Token
from opentoken.tokens.definitions.t5_token import T5Token

class TokenRegistry:
    """
    Loads all implementations of Token and returns their definitions.
    """
    @staticmethod
    def load_all_tokens() -> Dict[str, List[AttributeExpression]]:
        definitions: Dict[str, List[AttributeExpression]] = {}
        token_classes = [T1Token, T2Token, T3Token, T4Token, T5Token]
        for token_class in token_classes:
            token = token_class()
            definitions[token.get_identifier()] = token.get_definition()
        return definitions

File: lib/nodejs/opentoken/src/tokens/TokenRegistry.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { AttributeExpression } from '../attributes/AttributeExpression';
import { Token } from './Token';
import { T1Token } from './definitions/T1Token';
import { T2Token } from './definitions/T2Token';
import { T3Token } from './definitions/T3Token';
import { T4Token } from './definitions/T4Token';
import { T5Token } from './definitions/T5Token';

/**
 * Registry for all available tokens in the system.
 */
export class TokenRegistry {
  private constructor() {
    // Private constructor to prevent instantiation
  }

  /**
   * Load all available token definitions.
   */
  static loadAll(): Token[] {
    return [new T1Token(), new T2Token(), new T3Token(), new T4Token(), new T5Token()];
  }

  /**
   * Load all available token definitions as a map.
   * 
   * @returns a map where keys are token identifiers (e.g., "T1", "T2")
   *          and values are lists of AttributeExpression defining the token
   */
  static loadAllTokens(): Map<string, AttributeExpression[]> {
    const tokensMap = new Map<string, AttributeExpression[]>();
    const tokens = this.loadAll();
    for (const token of tokens) {
      tokensMap.set(token.getIdentifier(), token.getDefinition());
    }
    return tokensMap;
  }

  /**
   * Find a token by its ID.
   */
  static findById(id: string): Token | undefined {
    const tokens = this.loadAll();
    return tokens.find((token) => token.getIdentifier() === id);
  }
}
